CMakeLists.txt

```diff
@@ -484,42 +487,56 @@ endif ()
-  add_definitions(-DHAVE_INTEL_ATOMIC_PRIMITIVES)
-endif ()
-
-file(WRITE "${PROJECT_BINARY_DIR}/try_compile_solaris_atomic_ops.c"
-"		#include <atomic.h>
-		/* This requires Solaris Studio 12.2 or newer: */
-		#include <mbarrier.h>
-		void memory_barrier (void) { __machine_rw_barrier (); }
-		int atomic_add (volatile unsigned *i) { return atomic_add_int_nv (i, 1); }
-		void *atomic_ptr_cmpxchg (volatile void **target, void *cmp, void *newval) { return atomic_cas_ptr (target, cmp, newval); }
-		int main () { return 0; }
-")
-try_compile(HB_HAVE_SOLARIS_ATOMIC_OPS
-  ${PROJECT_BINARY_DIR}/try_compile_solaris_atomic_ops
-  ${PROJECT_BINARY_DIR}/try_compile_solaris_atomic_ops.c)
-if (HB_HAVE_SOLARIS_ATOMIC_OPS)
-  add_definitions(-DHAVE_SOLARIS_ATOMIC_OPS)
+if (NOT EMSCRIPTEN)
+  file(WRITE "${PROJECT_BINARY_DIR}/try_compile_intel_atomic_primitives.c"
+  "		void memory_barrier (void) { __sync_synchronize (); }
+      int atomic_add (int *i) { return __sync_fetch_and_add (i, 1); }
+      int mutex_trylock (int *m) { return __sync_lock_test_and_set (m, 1); }
+      void mutex_unlock (int *m) { __sync_lock_release (m); }
+      int main () { return 0; }
+  ")
+  try_compile(HB_HAVE_INTEL_ATOMIC_PRIMITIVES
+    ${PROJECT_BINARY_DIR}/try_compile_intel_atomic_primitives
+    ${PROJECT_BINARY_DIR}/try_compile_intel_atomic_primitives.c)
+  if (HB_HAVE_INTEL_ATOMIC_PRIMITIVES)
+    add_definitions(-DHAVE_INTEL_ATOMIC_PRIMITIVES)
+  endif ()
+
+  file(WRITE "${PROJECT_BINARY_DIR}/try_compile_solaris_atomic_ops.c"
+  "		#include <atomic.h>
+      /* This requires Solaris Studio 12.2 or newer: */
+      #include <mbarrier.h>
+      void memory_barrier (void) { __machine_rw_barrier (); }
+      int atomic_add (volatile unsigned *i) { return atomic_add_int_nv (i, 1); }
+      void *atomic_ptr_cmpxchg (volatile void **target, void *cmp, void *newval) { return atomic_cas_ptr (target, cmp, newval); }
+      int main () { return 0; }
+  ")
+  try_compile(HB_HAVE_SOLARIS_ATOMIC_OPS
+    ${PROJECT_BINARY_DIR}/try_compile_solaris_atomic_ops
+    ${PROJECT_BINARY_DIR}/try_compile_solaris_atomic_ops.c)
+  if (HB_HAVE_SOLARIS_ATOMIC_OPS)
+    add_definitions(-DHAVE_SOLARIS_ATOMIC_OPS)
+  endif ()
+else ()
+  # emscripten doesn't need multithreading support, so
+  add_definitions(-DHB_NO_MT)
 endif ()
 
 
 ## Define harfbuzz library
-add_library(harfbuzz ${project_sources} ${project_extra_sources} ${project_headers})
+if (NOT EMSCRIPTEN) # define emscripten as a executable on emscripten mode
+  add_library(harfbuzz ${project_sources} ${project_extra_sources} ${project_headers})
+else ()
+  add_executable(harfbuzz ${project_sources} ${project_extra_sources} ${project_headers})
+  add_definitions("-DHB_EXTERN=__attribute__((used))")
+  add_definitions(-DHB_NO_MT)
+  set (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -s MODULARIZE=1 \
+    -s 'EXTRA_EXPORTED_RUNTIME_METHODS=[\"ccall\", \"cwrap\"]' -s WASM=1 \
+    -fno-threadsafe-statics -fno-rtti -fno-exceptions -Oz")
```

Comment:
This line looks odd from Emscripten perspective: In CMake `CMAKE_CXX_FLAGS` specifies compilation time flags, and I think these flags will not be present at link time(?)

In Emscripten when `-s 'EXTRA_EXPORTED_RUNTIME_METHODS=[\"ccall\", \"cwrap\"]'` or `-s MODULARIZE=1` are present at compile time, they have no effect, instead these two are only link time flags.

The other flags, especially `-s WASM=1` apply at both compile and link time.

Specifying `-Oz` here seems slightly nonidiomatic use of CMake. For CMake, there's four canonical CMake build modes, Debug, RelWithDebInfo, Release and MinSizeRel. The CMake default iirc is that Debug is `-O0`, and Release and RelWithDebInfo are `-O3`, and MinSizeRel is `-Oz`. Specifying `-Oz` here will override the optimization mode even for debug and release builds. I wonder if this was intentional here to avoid some codegen bug, or performance issue, or similar? Perhaps debug builds still should not apply `-Oz`, but would be better with the default `-O0`?

Reply:
Honestly it was just an experiment, whole thing removed.